src/platform/datapath_epoll.c

```diff
@@ -1644,6 +1535,260 @@ CxPlatSocketContextSendComplete(
+    UNREFERENCED_PARAMETER(SegmentSize);
+    for (CMsg = CMSG_FIRSTHDR(Msg);
+        CMsg != NULL;
+        CMsg = CMSG_NXTHDR(Msg, CMsg)) {
+
+        if (CMsg->cmsg_level == IPPROTO_IPV6) {
+            if (CMsg->cmsg_type == IPV6_PKTINFO) {
+                struct in6_pktinfo* PktInfo6 = (struct in6_pktinfo*) CMSG_DATA(CMsg);
+                LocalAddr->Ip.sa_family = QUIC_ADDRESS_FAMILY_INET6;
+                LocalAddr->Ipv6.sin6_addr = PktInfo6->ipi6_addr;
+                LocalAddr->Ipv6.sin6_port = Socket->LocalAddress.Ipv6.sin6_port;
+                CxPlatConvertFromMappedV6(LocalAddr, LocalAddr);
+
+                LocalAddr->Ipv6.sin6_scope_id = PktInfo6->ipi6_ifindex;
+                FoundLocalAddr = TRUE;
+            } else if (CMsg->cmsg_type == IPV6_TCLASS) {
+                *ECN = *(uint8_t *)CMSG_DATA(CMsg);
+                FoundTOS = TRUE;
+            }
+        } else if (CMsg->cmsg_level == IPPROTO_IP) {
+            if (CMsg->cmsg_type == IP_PKTINFO) {
+                struct in_pktinfo* PktInfo = (struct in_pktinfo*)CMSG_DATA(CMsg);
+                LocalAddr->Ip.sa_family = QUIC_ADDRESS_FAMILY_INET;
+                LocalAddr->Ipv4.sin_addr = PktInfo->ipi_addr;
+                LocalAddr->Ipv4.sin_port = Socket->LocalAddress.Ipv6.sin6_port;
+                LocalAddr->Ipv6.sin6_scope_id = PktInfo->ipi_ifindex;
+                FoundLocalAddr = TRUE;
+            } else if (CMsg->cmsg_type == IP_TOS) {
+                *ECN = *(uint8_t *)CMSG_DATA(CMsg);
+                FoundTOS = TRUE;
+            }
+#ifdef UDP_GRO
+        } else if (CMsg->cmsg_type == IPPROTO_UDP) {
+            if (CMsg->cmsg_type == UDP_GRO) {
+                CXPLAT_DBG_ASSERT(*(uint16_t*)CMSG_DATA(CMsg) <= MAX_GRO_SEGMENT_SIZE);
+                *SegmentSize = (uint16_t)*(uint16_t*)CMSG_DATA(CMsg);
+            }
+#endif
+        }
+    }
+
+    // TODO Errors
+    return FoundLocalAddr && FoundTOS;
+}
+
+CXPLAT_DATAPATH_INTERNAL_RECV_BUFFER_CONTEXT*
+CxPlatDataPathDatagramToInternalDatagramContext(
+    _In_ CXPLAT_RECV_DATA* Datagram
+    );
+
+static
+QUIC_STATUS
+CxPlatSocketRecv(
+    _In_ CXPLAT_SOCKET_CONTEXT* SocketContext
+    )
+{
+    CXPLAT_DATAPATH_PROC_CONTEXT* ProcContext = SocketContext->ProcContext;
+    CXPLAT_DATAPATH* Datapath = SocketContext->Binding->Datapath;
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+
+    CXPLAT_DATAPATH_INTERNAL_RECV_CONTEXT* RecvContext =
+        CxPlatSocketAllocRecvContext(
+            ProcContext);
+
+    uint8_t* RecvPayload = ((uint8_t*)RecvContext) + SocketContext->Binding->Datapath->RecvPayloadOffset;
+
+    BOOLEAN UseURO = !!(SocketContext->Binding->Datapath->Features & CXPLAT_DATAPATH_FEATURE_RECV_COALESCING);
+    int BufferCount = UseURO ? 1 : NUMBER_OF_SEGMENTS_OR_BATCHES;
+    struct iovec RecvIov[NUMBER_OF_SEGMENTS_OR_BATCHES];
+    CXPLAT_RECV_MSG_CONTROL_BUFFER RecvMsgControl[NUMBER_OF_SEGMENTS_OR_BATCHES];
+    struct mmsghdr RecvMsgHdr[NUMBER_OF_SEGMENTS_OR_BATCHES];
```

Comment:
These 3 arrays total to about 15k, so I'll probably pull them out to a separate allocation. I'll do it on a per thread level though, so it will be a static allocation per thread.